lib/logstash/inputs/file.rb

```diff
@@ -181,27 +182,47 @@ def register
     if @start_position == "beginning"
       @tail_config[:start_new_files_at] = :beginning
     end
+
+    if (@run_with_identity = @codec.class.config_name == 'multiline')
+      @codec = LogStash::Codecs::IdentityMapCodec.new(@codec)
+    end
   end # def register
 
-  public
   def run(queue)
     @tail = FileWatch::Tail.new(@tail_config)
     @tail.logger = @logger
     @path.each { |path| @tail.tail(path) }
-
     @tail.subscribe do |path, line|
-      @logger.debug? && @logger.debug("Received line", :path => path, :text => line)
-      @codec.decode(line) do |event|
-        event["[@metadata][path]"] = path
-        event["host"] = @host if !event.include?("host")
-        event["path"] = path if !event.include?("path")
-        decorate(event)
-        queue << event
+      log_line_received(path, line)
+      if @run_with_identity
+        @codec.decode(line, path) do |event|
+          # path is the identity
+          # Note: this block is cached in the
+          # identity_map_codec for use when
+          # buffered lines are flushed.
+          queue << add_path_meta(event, path)
+        end
+      else
+        @codec.decode(line) do |event|
+          queue << add_path_meta(event, path)
+        end
       end
     end
   end # def run
 
-  public
+  def log_line_received(path, line)
+    return if !@logger.debug?
+    @logger.debug("Received line", :path => path, :text => line)
+  end
```

Comment:
It may be inlined by jruby optimization. But its real value is in decluttering the subscribe block so the next dev can 'get' the real purpose of the block.  I think its quite good practise to have many small methods which reveal their intent to the reader.